src/mechanics/IGA/NURBS.h

```diff
@@ -0,0 +1,478 @@
+    {
+        const std::array<int, TDimParameter> spanIdx = FindSpan(parameter);
+        Eigen::MatrixXd shapeFunctions = BasisFunctionsAndDerivativesRational(0, parameter);
+
+        Eigen::VectorXd coordinates(GetDimension());
+        coordinates.setZero(GetDimension());
+
+        Eigen::VectorXd cpCoords = GetControlPointsElement(spanIdx);
+
+        std::cout << "shapeFunctions: " << shapeFunctions << std::endl;
+
+        int numCPs = GetNumControlPointsElement();
+
+        assert(numCPs == shapeFunctions.rows());
+
+        int dim = GetDimension();
+
+        for(int i = 0; i < dim; i++)
+            for(int j = 0; j < numCPs; j++)
+                coordinates(i) += shapeFunctions(j,0) * cpCoords(dim*j+i);
+
+        return coordinates;
+    }
+
+    /** Basis Functions **/
+
+    static int FindSpan(double rParameter, int rDegree, const std::vector<double> &rKnots)
+    {
+        int size = rKnots.size();
+        int iterations = 0;
+
+        if(rParameter < rKnots[0] || rParameter > rKnots[size - 1])
+            throw new std::string("The parameter is out of the range of the knot vector.");
+
+        int numBasisFuns = size - rDegree - 1;
+        if(rParameter == rKnots[numBasisFuns]) return numBasisFuns-1;
+
+        int low  = rDegree;
+        int high = numBasisFuns;
+        int mid  = (low + high)/2;
+
+        while(rParameter < rKnots[mid] || rParameter >= rKnots[mid+1])
+        {
+            if(rParameter < rKnots[mid]) high = mid;
+            else                         low = mid;
+
+            mid = (low + high)/2;
+            iterations++;
+            if(iterations > size)
+                throw new std::string("The maximum number of iterations for finding the span exceeded.");
+        }
+        return mid;
+    }
+
+    static Eigen::MatrixXd BasisFunctionsAndDerivatives(int der, double parameter, int spanIdx, int rDegree, const std::vector<double> &rKnots)
+    {
+        Eigen::MatrixXd ndu(rDegree+1, rDegree+1);
+
+        ndu(0,0) = 1.0;
+
+        Eigen::VectorXd left(rDegree + 1);
+        Eigen::VectorXd right(rDegree + 1);
+
+        for (int j = 1; j <= rDegree; j++)
+        {
+            left[j]  = parameter - rKnots[spanIdx + 1 - j];
+            right[j] = rKnots[spanIdx + j] - parameter;
+            double saved = 0.;
+            for(int r = 0; r < j; r++)
+            {
+                ndu(j,r) = right[r+1] + left[j-r]; // lower triange (knot differences)
+                double temp = ndu(r, j-1)/ndu(j,r);
+                ndu(r,j) = saved + right[r+1]*temp; // upper triangle
+                saved = left[j-r]*temp;
+            }
+            ndu(j,j) = saved;
+        }
+
+        Eigen::MatrixXd ders(der+1, rDegree+1);
+
+        for(int j = 0; j <= rDegree; j++) ders(0,j) = ndu(j,rDegree);
+
+        Eigen::MatrixXd a(2, rDegree+1);
+        for(int r = 0; r<=rDegree; r++)
+        {
+            int s1 = 0;
```

Comment:
Is there a book/paper that you could refer to, where one would be able to see the corresponding theory and immediately know what `s1, s2, a, j, k, rk, pk, d, j1, j2` are? If so, please mention it in a comment.